File: 백준/Silver/9012. 괄호/괄호.py
import sys

N = int(sys.stdin.readline())

def check(str):
    stack = []
    for s in str:
        if s=='(':
            stack.append(0)
        elif s==')':
            if(len(stack)==0):
                return False
            else:
                stack.pop()
    if len(stack)==0:
        return True
    else:
        return False

for _ in range(N):
    str = sys.stdin.readline()
    if check(str):
        print('YES')
    else:
        print('NO')

File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>
using namespace std;

bool check(const string& str) {
    stack<int> stack;
    for (char s : str) {
        if (s == '(') {
            stack.push(0);
        } else if (s == ')') {
            if (stack.empty()) {
                return false;
            } else {
                stack.pop();
            }
        }
    }
    return stack.empty();  // 비어있다면 true 아니면 false
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    int T;
    string str;
    cin >> T;
    while (T--){
        cin >> str;
        if (check(str)) cout << "YES" << '\n';
        else cout << "NO" << '\n';
    }
    
    return 0;
}